pallets/phala-world/src/lib.rs

```diff
@@ -0,0 +1,814 @@
+						},
+						_ => T::NormalEggPrice::get(),
+					};
+					ensure!(egg_price != T::NormalEggPrice::get(), Error::<T>::InvalidPurchase);
+					let nft_id = pallet_rmrk_core::NextNftId::<T>::get(EGGS_COLLECTION_ID);
+					// Transfer the amount for the rare Egg NFT then mint the egg
+					<T as pallet::Config>::Currency::transfer(
+						&sender,
+						&overlord,
+						egg_price,
+						ExistenceRequirement::KeepAlive
+					)?;
+					// Mint Egg and transfer Egg to new owner
+					pallet_rmrk_core::Pallet::<T>::mint_nft(
+						Origin::<T>::Signed(overlord.clone()).into(),
+						sender.clone(),
+						EGGS_COLLECTION_ID,
+						None,
+						None,
+						metadata,
+					)?;
+					// Add EggInfo to storage
+					let egg = EggInfo {
+						egg_type,
+						race,
+						career,
+						start_hatching: 0,
+						hatching_duration: 0,
+					};
+					Eggs::<T>::insert(EGGS_COLLECTION_ID, nft_id, egg);
+
+					Self::deposit_event(Event::RareEggPurchased {
+						collection_id: EGGS_COLLECTION_ID,
+						nft_id,
+						owner: sender,
+					});
+
+					Ok(())
+				}
+			}
+		}
+
+		/// Users can pre-order an egg. This will enable users that are whitelisted to be
+		/// added to the queue of users that can claim eggs. Those that come after the whitelist
+		/// pre-sale will be able to win the chance to acquire an egg based on their choice of
+		/// race and career as they will have a limited quantity.
+		///
+		/// Parameters:
+		/// - origin: The origin of the extrinsic preordering the egg
+		/// - race: The race that the user has chosen (limited # of races)
+		/// - career: The career that the user has chosen (limited # of careers)
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(1,1))]
+		#[transactional]
+		pub fn preorder_egg(
+			origin: OriginFor<T>,
+			race: RaceType,
+			career: CareerType,
+		) -> DispatchResult {
+			ensure!(CanPreorderEggs::<T>::get(), Error::<T>::PreorderEggNotAvailable);
+			let sender = ensure_signed(origin)?;
+
+			// Reserve currency for the preorder at the Normal egg price
+			<T as pallet::Config>::Currency::reserve(&sender, T::NormalEggPrice::get())?;
+
+			let preorder = PreorderInfo {
+				owner: sender.clone(),
+				race,
+				career,
+			};
+			<Preorders<T>>::get().push(preorder.into());
+
+			Self::deposit_event(Event::EggPreordered {
+				owner: sender,
+			});
+
+			Ok(())
+		}
+
+		/// This is an admin only function that will be called to do a bulk minting of all egg
+		/// owners that made selected a race and career that was available based on the quantity
+		/// available. Those that did not win an egg will have to claim their refund
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(1,1))]
+		#[transactional]
+		pub fn mint_eggs(
+			origin: OriginFor<T>,
+		) -> DispatchResult {
+			// Ensure Overlord account makes call
+			let sender = ensure_signed(origin)?;
+			ensure!(Self::overlord().map_or(false, |k| sender == k), Error::<T>::RequireOverlordAccount);
```

Comment:
I suggest to make a function `fn ensure_overload(&self, sender: &T::AccountId) -> DispatchResult` since it will be used multiple times